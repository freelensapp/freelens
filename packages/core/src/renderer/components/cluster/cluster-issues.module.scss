/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

 .ClusterIssues {
  min-height: 350px;
  position: relative;
  grid-area: issues;
  padding: calc(var(--margin) * 2) 0;
  background: var(--contentColor);

  &:global(.OnlyClusterIssues) {
    grid-row: row1-start / row2-end;
  }

  @media (max-width: 1150px) {
    grid-area: unset;

    &:global(.OnlyClusterIssues) {
      grid-row: auto;
    }
  }

  .SubHeader {
    padding-top: 0;
    padding-bottom: 0;

    :global(.Icon) {
      font-size: 130%;
      color: var(--colorError);
    }
  }

  .message {
    white-space: nowrap;
    text-overflow: ellipsis;
    flex-grow: 3;
  }

  .object {
    white-space: nowrap;
    text-overflow: ellipsis;
    flex-grow: 2;
  }

   .noIssues {
     .Icon {
       color: white;
     }

     .title {
       font-size: large;
       color: var(--textColorAccent);
       font-weight: bold;
     }
   }
}
