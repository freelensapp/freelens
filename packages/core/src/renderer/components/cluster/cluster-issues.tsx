/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import styles from "./cluster-issues.module.scss";

import { Icon } from "@freelensapp/icon";
import { Spinner } from "@freelensapp/spinner";
import { cssNames, prevDefault } from "@freelensapp/utilities";
import { withInjectables } from "@ogre-tools/injectable-react";
import type { IComputedValue } from "mobx";
import { computed, makeObservable } from "mobx";
import { observer } from "mobx-react";
import React from "react";
import type { ApiManager } from "../../../common/k8s-api/api-manager";
import apiManagerInjectable from "../../../common/k8s-api/api-manager/manager.injectable";
import type { PageParam } from "../../navigation/page-param";
import activeThemeInjectable from "../../themes/active.injectable";
import type { LensTheme } from "../../themes/lens-theme";
import type { EventStore } from "../events/store";
import eventStoreInjectable from "../events/store.injectable";
import kubeSelectedUrlParamInjectable from "../kube-detail-params/kube-selected-url.injectable";
import type { ToggleKubeDetailsPane } from "../kube-detail-params/toggle-details.injectable";
import toggleKubeDetailsPaneInjectable from "../kube-detail-params/toggle-details.injectable";
import { KubeObjectAge } from "../kube-object/age";
import { SubHeader } from "../layout/sub-header";
import type { NodeStore } from "../nodes/store";
import nodeStoreInjectable from "../nodes/store.injectable";
import { Table, TableCell, TableHead, TableRow } from "../table";

export interface ClusterIssuesProps {
  className?: string;
}

interface Warning {
  getId: () => string;
  getName: () => string;
  kind: string;
  message: string | undefined;
  selfLink: string;
  renderAge: () => React.ReactElement;
  ageMs: number;
}

enum sortBy {
  type = "type",
  object = "object",
  age = "age",
}

interface Dependencies {
  activeTheme: IComputedValue<LensTheme>;
  nodeStore: NodeStore;
  eventStore: EventStore;
  apiManager: ApiManager;
  kubeSelectedUrlParam: PageParam<string>;
  toggleKubeDetailsPane: ToggleKubeDetailsPane;
}

@observer
class NonInjectedClusterIssues extends React.Component<ClusterIssuesProps & Dependencies> {
  constructor(props: ClusterIssuesProps & Dependencies) {
    super(props);
    makeObservable(this);
  }

  @computed get warnings(): Warning[] {
    return [
      ...this.props.nodeStore.items.flatMap((node) =>
        node.getWarningConditions().map(({ message }) => ({
          selfLink: node.selfLink,
          getId: () => node.getId(),
          getName: () => node.getName(),
          kind: node.kind,
          message,
          renderAge: () => <KubeObjectAge key="age" object={node} />,
          ageMs: -node.getCreationTimestamp(),
        })),
      ),
      ...this.props.eventStore.getWarnings().map((warning) => ({
        getId: () => warning.involvedObject.uid,
        getName: () => warning.involvedObject.name,
        renderAge: () => <KubeObjectAge key="age" object={warning} />,
        ageMs: -warning.getCreationTimestamp(),
        message: warning.message,
        kind: warning.kind,
        selfLink: this.props.apiManager.lookupApiLink(warning.involvedObject, warning),
      })),
    ];
  }

  getTableRow = (uid: string) => {
    const { warnings } = this;
    const { kubeSelectedUrlParam, toggleKubeDetailsPane: toggleDetails } = this.props;
    const warning = warnings.find((warn) => warn.getId() == uid);

    if (!warning) {
      return undefined;
    }

    const { getId, getName, message, kind, selfLink, renderAge } = warning;

    return (
      <TableRow
        key={getId()}
        sortItem={warning}
        selected={selfLink === kubeSelectedUrlParam.get()}
        onClick={prevDefault(() => toggleDetails(selfLink))}
      >
        <TableCell className={styles.message}>{message ?? "<unknown>"}</TableCell>
        <TableCell className={styles.object}>{getName()}</TableCell>
        <TableCell className="kind">{kind}</TableCell>
        <TableCell className="age">{renderAge()}</TableCell>
      </TableRow>
    );
  };

  renderContent() {
    const { warnings } = this;

    if (!this.props.eventStore.isLoaded) {
      return <Spinner center />;
    }

    if (!warnings.length) {
      return (
        <div className={cssNames(styles.noIssues, "flex column box grow gaps align-center justify-center")}>
          <Icon className={styles.Icon} material="check" big sticker />
          <p className={styles.title}>No issues found</p>
          <p>Everything is fine in the Cluster</p>
        </div>
      );
    }

    return (
      <>
        <SubHeader className={styles.SubHeader}>
          <Icon material="error_outline" />
          {` Warnings: ${warnings.length}`}
        </SubHeader>
        <Table
          tableId="cluster_issues"
          items={warnings}
          virtual
          selectable
          sortable={{
            [sortBy.type]: (warning) => warning.kind,
            [sortBy.object]: (warning) => warning.getName(),
            [sortBy.age]: (warning) => warning.ageMs,
          }}
          sortByDefault={{ sortBy: sortBy.object, orderBy: "asc" }}
          sortSyncWithUrl={false}
          getTableRow={this.getTableRow}
          className={cssNames("box grow", this.props.activeTheme.get().type)}
        >
          <TableHead nowrap>
            <TableCell className="message">Message</TableCell>
            <TableCell className="object" sortBy={sortBy.object}>
              Object
            </TableCell>
            <TableCell className="kind" sortBy={sortBy.type}>
              Type
            </TableCell>
            <TableCell className="timestamp" sortBy={sortBy.age}>
              Age
            </TableCell>
          </TableHead>
        </Table>
      </>
    );
  }

  render() {
    return (
      <div className={cssNames(styles.ClusterIssues, "flex column", this.props.className)}>{this.renderContent()}</div>
    );
  }
}

export const ClusterIssues = withInjectables<Dependencies, ClusterIssuesProps>(NonInjectedClusterIssues, {
  getProps: (di, props) => ({
    ...props,
    activeTheme: di.inject(activeThemeInjectable),
    apiManager: di.inject(apiManagerInjectable),
    eventStore: di.inject(eventStoreInjectable),
    nodeStore: di.inject(nodeStoreInjectable),
    kubeSelectedUrlParam: di.inject(kubeSelectedUrlParamInjectable),
    toggleKubeDetailsPane: di.inject(toggleKubeDetailsPaneInjectable),
  }),
});
