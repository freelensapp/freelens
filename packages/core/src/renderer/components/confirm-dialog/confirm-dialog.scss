/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

@use "../vars" as *;

.ConfirmDialog {
  $spacing: $padding * 3;

  &.warning {
    .Button.ok {
      background: var(--colorSoftError);
    }
  }

  > .box {
    max-width: 50vw;
    min-width: 45 * $unit;
    background-color: white;
    border-radius: $radius;
    line-height: 1.5;
  }

  .confirm-content {
    padding: $spacing;
    display: flex;
    align-items: center;
    word-break: break-word;

    > .Icon {
      margin-left: inherit;
      margin-right: $margin * 2;
      color: var(--colorSoftError);
    }

    > :not(.Icon) {
      flex: 1;
    }

    .warning {
      font-size: small;
      margin-top: $margin;
    }
  }

  .confirm-buttons {
    background: #f4f4f4;
    padding: $padding * 2.5;
    display: flex;
    justify-content: flex-end;
    border-bottom-left-radius: $radius;
    border-bottom-right-radius: $radius;

    > * {
      margin-left: $margin
    }
  }
}
