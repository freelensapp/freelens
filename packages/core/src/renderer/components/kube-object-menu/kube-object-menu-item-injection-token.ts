/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { KubeObject } from "@freelensapp/kube-object";
import { getInjectionToken } from "@ogre-tools/injectable";
import type { IComputedValue } from "mobx";
import type React from "react";
import type { KubeObjectMenuProps } from "./kube-object-menu";

export type KubeObjectMenuItemComponent = React.ElementType<KubeObjectMenuProps<KubeObject>>;

export interface KubeObjectMenuItem {
  kind: string;
  apiVersions: string[];
  enabled: IComputedValue<boolean>;
  Component: KubeObjectMenuItemComponent;
  orderNumber: number;
}

export const kubeObjectMenuItemInjectionToken = getInjectionToken<KubeObjectMenuItem>({
  id: "kube-object-menu-item-injection-token",
});
