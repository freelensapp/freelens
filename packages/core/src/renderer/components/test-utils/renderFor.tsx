/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { DiContainer } from "@ogre-tools/injectable";
import { DiContextProvider } from "@ogre-tools/injectable-react";
import type { RenderResult } from "@testing-library/react";
import { render as testingLibraryRender } from "@testing-library/react";
import React from "react";

export type DiRender = (ui: React.ReactElement) => RenderResult;

type DiRenderFor = (di: DiContainer) => DiRender;

export const renderFor: DiRenderFor = (di) => (ui) => {
  const result = testingLibraryRender(<DiContextProvider value={{ di }}>{ui}</DiContextProvider>);

  return {
    ...result,

    rerender: (ui: React.ReactElement) => result.rerender(<DiContextProvider value={{ di }}>{ui}</DiContextProvider>),
  };
};
