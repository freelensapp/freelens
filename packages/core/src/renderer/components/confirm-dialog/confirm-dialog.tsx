/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./confirm-dialog.scss";

import type { ButtonProps } from "@freelensapp/button";
import { Button } from "@freelensapp/button";
import { Icon } from "@freelensapp/icon";
import type { ShowNotification } from "@freelensapp/notifications";
import { showErrorNotificationInjectable } from "@freelensapp/notifications";
import type { StrictReactNode } from "@freelensapp/utilities";
import { cssNames, noop, prevDefault } from "@freelensapp/utilities";
import { withInjectables } from "@ogre-tools/injectable-react";
import type { IObservableValue } from "mobx";
import { computed, makeObservable, observable } from "mobx";
import { observer } from "mobx-react";
import React from "react";
import type { DialogProps } from "../dialog";
import { Dialog } from "../dialog";
import confirmDialogStateInjectable from "./state.injectable";

export interface ConfirmDialogProps extends Partial<DialogProps> {}

export interface ConfirmDialogParams extends ConfirmDialogBooleanParams {
  ok?: () => any | Promise<any>;
  cancel?: () => any | Promise<any>;
}

export interface ConfirmDialogBooleanParams {
  labelOk?: StrictReactNode;
  labelCancel?: StrictReactNode;
  message: StrictReactNode;
  icon?: StrictReactNode;
  okButtonProps?: Partial<ButtonProps>;
  cancelButtonProps?: Partial<ButtonProps>;
}

interface Dependencies {
  state: IObservableValue<ConfirmDialogParams | undefined>;
  showErrorNotification: ShowNotification;
}

const defaultParams = {
  ok: noop,
  cancel: noop,
  labelOk: "Ok",
  labelCancel: "Cancel",
  icon: <Icon big material="warning" />,
};

@observer
class NonInjectedConfirmDialog extends React.Component<ConfirmDialogProps & Dependencies> {
  @observable isSaving = false;

  constructor(props: ConfirmDialogProps & Dependencies) {
    super(props);
    makeObservable(this);
  }

  @computed
  get params() {
    return Object.assign({}, defaultParams, this.props.state.get() ?? ({} as ConfirmDialogParams));
  }

  ok = async () => {
    try {
      this.isSaving = true;
      await (async () => this.params.ok())();
    } catch (error) {
      this.props.showErrorNotification(
        <>
          <p>Confirmation action failed:</p>
          <p>
            {error instanceof Error
              ? error.message
              : typeof error === "string"
                ? error
                : "Unknown error occurred while ok-ing"}
          </p>
        </>,
      );
    } finally {
      this.isSaving = false;
      this.props.state.set(undefined);
    }
  };

  onClose = () => {
    this.isSaving = false;
  };

  close = async () => {
    try {
      await Promise.resolve(this.params.cancel());
    } catch (error) {
      this.props.showErrorNotification(
        <>
          <p>Cancelling action failed:</p>
          <p>
            {error instanceof Error
              ? error.message
              : typeof error === "string"
                ? error
                : "Unknown error occurred while cancelling"}
          </p>
        </>,
      );
    } finally {
      this.isSaving = false;
      this.props.state.set(undefined);
    }
  };

  render() {
    const { state, className, ...dialogProps } = this.props;
    const isOpen = Boolean(state.get());
    const { icon, labelOk, labelCancel, message, okButtonProps = {}, cancelButtonProps = {} } = this.params;

    return (
      <Dialog
        {...dialogProps}
        className={cssNames("ConfirmDialog", className)}
        isOpen={isOpen}
        onClose={this.onClose}
        close={this.close}
        {...(isOpen ? { "data-testid": "confirmation-dialog" } : {})}
      >
        <div className="confirm-content">
          {icon} {message}
        </div>
        <div className="confirm-buttons">
          <Button
            plain
            className="cancel"
            label={labelCancel}
            onClick={prevDefault(this.close)}
            {...cancelButtonProps}
          />
          <Button
            autoFocus
            primary
            className="ok"
            label={labelOk}
            onClick={prevDefault(this.ok)}
            waiting={this.isSaving}
            data-testid="confirm"
            {...okButtonProps}
          />
        </div>
      </Dialog>
    );
  }
}

export const ConfirmDialog = withInjectables<Dependencies, ConfirmDialogProps>(NonInjectedConfirmDialog, {
  getProps: (di, props) => ({
    ...props,
    state: di.inject(confirmDialogStateInjectable),
    showErrorNotification: di.inject(showErrorNotificationInjectable),
  }),
});
