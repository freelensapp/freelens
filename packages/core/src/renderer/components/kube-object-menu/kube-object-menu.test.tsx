/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import type { RenderResult } from "@testing-library/react";
import { screen, waitFor } from "@testing-library/react";
import React from "react";
import "@testing-library/jest-dom";
import type { AsyncFnMock } from "@async-fn/jest";
import asyncFn from "@async-fn/jest";
import { KubeObject } from "@freelensapp/kube-object";
import { getInjectable } from "@ogre-tools/injectable";
import type { DiContainer } from "@ogre-tools/injectable";
import type { UserEvent } from "@testing-library/user-event";
import userEvent from "@testing-library/user-event";
import { computed, runInAction } from "mobx";
import directoryForTempInjectable from "../../../common/app-paths/directory-for-temp/directory-for-temp.injectable";
import directoryForUserDataInjectable from "../../../common/app-paths/directory-for-user-data/directory-for-user-data.injectable";
import { Cluster } from "../../../common/cluster/cluster";
import type { ApiManager } from "../../../common/k8s-api/api-manager";
import apiManagerInjectable from "../../../common/k8s-api/api-manager/manager.injectable";
import clustersStateInjectable from "../../../features/cluster/storage/common/state.injectable";
import activeEntityIdInjectable from "../../api/catalog/entity/active-entity-id.injectable";
import activeEntityInternalClusterInjectable from "../../api/catalog/entity/get-active-cluster-entity.injectable";
import hostedClusterIdInjectable from "../../cluster-frame-context/hosted-cluster-id.injectable";
import { getDiForUnitTesting } from "../../getDiForUnitTesting";
import { ConfirmDialog } from "../confirm-dialog";
import createEditResourceTabInjectable from "../dock/edit-resource/edit-resource-tab.injectable";
import hideDetailsInjectable from "../kube-detail-params/hide-details.injectable";
import type { DiRender } from "../test-utils/renderFor";
import { renderFor } from "../test-utils/renderFor";
import { KubeObjectMenu } from "./index";
import { kubeObjectMenuItemInjectionToken } from "./kube-object-menu-item-injection-token";

// TODO: make `animated={false}` not required to make tests deterministic
describe("kube-object-menu", () => {
  let di: DiContainer;
  let render: DiRender;
  let user: UserEvent;

  beforeEach(() => {
    di = getDiForUnitTesting();

    di.override(hostedClusterIdInjectable, () => "some-cluster-id");
    di.override(directoryForUserDataInjectable, () => "/some-directory-for-user-data");
    di.override(directoryForTempInjectable, () => "/some-directory-for-temp");

    di.inject(clustersStateInjectable).set(
      "some-cluster-id",
      new Cluster({
        id: "some-cluster-id",
        contextName: "some-context-name",
        kubeConfigPath: "/some-path-to-a-kubeconfig",
      }),
    );
    di.override(activeEntityIdInjectable, () => computed(() => "some-cluster-id"));

    runInAction(() => {
      di.register(someMenuItemInjectable, someOtherMenuItemInjectable, someAnotherMenuItemInjectable);
    });

    render = renderFor(di);

    di.override(
      apiManagerInjectable,
      () =>
        ({
          getStore: (api: any) => void api,
        }) as ApiManager,
    );

    di.override(hideDetailsInjectable, () => () => {});

    di.override(createEditResourceTabInjectable, () => () => "irrelevant");

    user = userEvent.setup();
  });

  it("given no cluster, does not crash", () => {
    di.override(activeEntityInternalClusterInjectable, () => computed(() => undefined));

    expect(() => {
      render(<KubeObjectMenu object={null as never} toolbar={true} />);
    }).not.toThrow();
  });

  it("given no kube object, renders", () => {
    const { baseElement } = render(<KubeObjectMenu object={null as never} toolbar={true} />);

    expect(baseElement).toMatchSnapshot();
  });

  describe("given kube object", () => {
    let result: RenderResult;
    let removeActionMock: AsyncFnMock<() => void>;

    beforeEach(async () => {
      const objectStub = KubeObject.create({
        apiVersion: "some-api-version",
        kind: "some-kind",
        metadata: {
          uid: "some-uid",
          name: "some-name",
          resourceVersion: "some-resource-version",
          namespace: "some-namespace",
          selfLink: "/foo",
        },
      });

      removeActionMock = asyncFn();
      result = render(
        <div>
          <ConfirmDialog animated={false} />

          <KubeObjectMenu object={objectStub} toolbar={true} removeAction={removeActionMock} />
        </div>,
      );
    });

    it("renders", () => {
      expect(result.baseElement).toMatchSnapshot();
    });

    it("does not open a confirmation dialog yet", () => {
      expect(screen.queryByTestId("confirmation-dialog")).toBeNull();
    });

    describe("when rerendered with different kube object", () => {
      beforeEach(() => {
        const newObjectStub = KubeObject.create({
          apiVersion: "some-other-api-version",
          kind: "some-other-kind",
          metadata: {
            uid: "some-other-uid",
            name: "some-other-name",
            resourceVersion: "some-other-resource-version",
            namespace: "some-other-namespace",
            selfLink: "/some-other-api-version/some-other-kind/some-other-namespace/some-other-name",
          },
        });

        result.rerender(
          <div>
            <ConfirmDialog animated={false} />

            <KubeObjectMenu object={newObjectStub} toolbar={true} removeAction={removeActionMock} />
          </div>,
        );
      });

      it("renders", () => {
        expect(result.baseElement).toMatchSnapshot();
      });

      describe("when removing new kube object", () => {
        beforeEach(async () => {
          await user.click(
            await screen.findByTestId(
              "menu-action-delete-for-/some-other-api-version/some-other-kind/some-other-namespace/some-other-name",
            ),
          );
        });

        it("renders", async () => {
          await screen.findByTestId("confirmation-dialog");
          expect(result.baseElement).toMatchSnapshot();
        });
      });
    });

    describe("when removing kube object", () => {
      beforeEach(async () => {
        await user.click(await screen.findByTestId("menu-action-delete-for-/foo"));
      });

      it("renders", async () => {
        await screen.findByTestId("confirmation-dialog");
        expect(result.baseElement).toMatchSnapshot();
      });

      describe("when remove is confirmed", () => {
        beforeEach(async () => {
          const confirmRemovalButton = await screen.findByTestId("confirm");

          await user.click(confirmRemovalButton);
        });

        it("calls for removal of the kube object", () => {
          expect(removeActionMock).toHaveBeenCalledWith();
        });

        it("does not close the confirmation dialog yet", async () => {
          await screen.findByTestId("confirmation-dialog");
        });

        it("when removal resolves, closes the confirmation dialog", async () => {
          await removeActionMock.resolve();
          await waitFor(() => {
            expect(screen.queryByTestId("confirmation-dialog")).toBeNull();
          });
        });
      });
    });
  });

  describe("given kube object with namespace", () => {
    let baseElement: Element;

    beforeEach(async () => {
      const objectStub = KubeObject.create({
        apiVersion: "some-api-version",
        kind: "some-kind",
        metadata: {
          uid: "some-uid",
          name: "some-name",
          resourceVersion: "some-resource-version",
          namespace: "some-namespace",
          selfLink: "/foo",
        },
      });

      ({ baseElement } = render(
        <div>
          <ConfirmDialog animated={false} />

          <KubeObjectMenu object={objectStub} toolbar={true} removeAction={() => {}} />
        </div>,
      ));
    });

    it("when removing kube object, renders confirmation dialog with namespace", async () => {
      const menuItem = await screen.findByTestId("menu-action-delete-for-/foo");

      await user.click(menuItem);

      expect(baseElement).toMatchSnapshot();
    });
  });

  describe("given kube object without namespace", () => {
    let baseElement: Element;

    beforeEach(async () => {
      const objectStub = KubeObject.create({
        apiVersion: "some-api-version",
        kind: "some-kind",
        metadata: {
          uid: "some-uid",
          name: "some-name",
          resourceVersion: "some-resource-version",
          namespace: undefined,
          selfLink: "/foo",
        },
      });

      ({ baseElement } = render(
        <div>
          <ConfirmDialog animated={false} />

          <KubeObjectMenu object={objectStub} toolbar={true} removeAction={() => {}} />
        </div>,
      ));
    });

    it("when removing kube object, renders confirmation dialog without namespace", async () => {
      const menuItem = await screen.findByTestId("menu-action-delete-for-/foo");

      await user.click(menuItem);

      expect(baseElement).toMatchSnapshot();
    });
  });
});

const MenuItemComponent = () => <li>Some menu item</li>;

const someMenuItemInjectable = getInjectable({
  id: "some-menu-item",

  instantiate: () => ({
    apiVersions: ["some-api-version"],
    kind: "some-kind",
    Component: MenuItemComponent,
    enabled: computed(() => true),
    orderNumber: 1,
  }),

  injectionToken: kubeObjectMenuItemInjectionToken,
});

const someOtherMenuItemInjectable = getInjectable({
  id: "some-other-menu-item",

  instantiate: () => ({
    apiVersions: ["some-unrelated-api-version"],
    kind: "some-kind",
    Component: MenuItemComponent,
    enabled: computed(() => true),
    orderNumber: 1,
  }),

  injectionToken: kubeObjectMenuItemInjectionToken,
});

const someAnotherMenuItemInjectable = getInjectable({
  id: "some-another-menu-item",

  instantiate: () => ({
    apiVersions: ["some-api-version"],
    kind: "some-unrelated-kind",
    Component: MenuItemComponent,
    enabled: computed(() => true),
    orderNumber: 1,
  }),

  injectionToken: kubeObjectMenuItemInjectionToken,
});
