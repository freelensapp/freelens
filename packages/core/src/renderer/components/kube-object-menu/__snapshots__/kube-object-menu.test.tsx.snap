// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kube-object-menu given kube object renders 1`] = `
<body>
  <div>
    <div>
      <ul
        class="Menu MenuActions flex KubeObjectMenu toolbar gaps right bottom"
        data-testid="menu-actions-for-kube-object-menu-for-some-uid"
        id="menu-actions-for-kube-object-menu-for-some-uid"
      >
        <li>
          Some menu item
        </li>
        <li
          class="MenuItem"
          data-testid="menu-action-delete-for-/foo"
          tabindex="0"
        >
          <i
            class="Icon material interactive focusable"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="delete"
            >
              delete
            </span>
          </i>
          <div>
            Delete
          </div>
          <span
            class="title"
          >
            Delete
          </span>
        </li>
      </ul>
    </div>
  </div>
</body>
`;

exports[`kube-object-menu given kube object when removing kube object renders 1`] = `
<body>
  <div>
    <div>
      <ul
        class="Menu MenuActions flex KubeObjectMenu toolbar gaps right bottom"
        data-testid="menu-actions-for-kube-object-menu-for-some-uid"
        id="menu-actions-for-kube-object-menu-for-some-uid"
      >
        <li>
          Some menu item
        </li>
        <li
          class="MenuItem"
          data-testid="menu-action-delete-for-/foo"
          tabindex="0"
        >
          <i
            class="Icon material interactive focusable"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="delete"
            >
              delete
            </span>
          </i>
          <div>
            Delete
          </div>
          <span
            class="title"
          >
            Delete
          </span>
        </li>
      </ul>
    </div>
  </div>
  <div
    class="Dialog flex center ConfirmDialog modal"
    data-testid="confirmation-dialog"
  >
    <div
      class="box"
    >
      <div
        class="confirm-content"
      >
        <i
          class="Icon material focusable big"
        >
          <span
            class="icon"
            data-icon-name="warning"
          >
            warning
          </span>
        </i>
         
        <p>
          Remove some-kind 
          <b>
            some-namespace/some-name
          </b>
           from 
          <b>
            some-context-name
          </b>
          ?
        </p>
      </div>
      <div
        class="confirm-buttons"
      >
        <button
          class="Button cancel plain"
          type="button"
        >
          Cancel
        </button>
        <button
          class="Button ok primary"
          data-testid="confirm"
          data-waiting="false"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`kube-object-menu given kube object when rerendered with different kube object renders 1`] = `
<body>
  <div>
    <div>
      <ul
        class="Menu MenuActions flex KubeObjectMenu toolbar gaps right bottom"
        data-testid="menu-actions-for-kube-object-menu-for-some-other-uid"
        id="menu-actions-for-kube-object-menu-for-some-other-uid"
      >
        <li
          class="MenuItem"
          data-testid="menu-action-delete-for-/some-other-api-version/some-other-kind/some-other-namespace/some-other-name"
          tabindex="0"
        >
          <i
            class="Icon material interactive focusable"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="delete"
            >
              delete
            </span>
          </i>
          <div>
            Delete
          </div>
          <span
            class="title"
          >
            Delete
          </span>
        </li>
      </ul>
    </div>
  </div>
</body>
`;

exports[`kube-object-menu given kube object when rerendered with different kube object when removing new kube object renders 1`] = `
<body>
  <div>
    <div>
      <ul
        class="Menu MenuActions flex KubeObjectMenu toolbar gaps right bottom"
        data-testid="menu-actions-for-kube-object-menu-for-some-other-uid"
        id="menu-actions-for-kube-object-menu-for-some-other-uid"
      >
        <li
          class="MenuItem"
          data-testid="menu-action-delete-for-/some-other-api-version/some-other-kind/some-other-namespace/some-other-name"
          tabindex="0"
        >
          <i
            class="Icon material interactive focusable"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="delete"
            >
              delete
            </span>
          </i>
          <div>
            Delete
          </div>
          <span
            class="title"
          >
            Delete
          </span>
        </li>
      </ul>
    </div>
  </div>
  <div
    class="Dialog flex center ConfirmDialog modal"
    data-testid="confirmation-dialog"
  >
    <div
      class="box"
    >
      <div
        class="confirm-content"
      >
        <i
          class="Icon material focusable big"
        >
          <span
            class="icon"
            data-icon-name="warning"
          >
            warning
          </span>
        </i>
         
        <p>
          Remove some-other-kind 
          <b>
            some-other-namespace/some-other-name
          </b>
           from 
          <b>
            some-context-name
          </b>
          ?
        </p>
      </div>
      <div
        class="confirm-buttons"
      >
        <button
          class="Button cancel plain"
          type="button"
        >
          Cancel
        </button>
        <button
          class="Button ok primary"
          data-testid="confirm"
          data-waiting="false"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`kube-object-menu given kube object with namespace when removing kube object, renders confirmation dialog with namespace 1`] = `
<body>
  <div>
    <div>
      <ul
        class="Menu MenuActions flex KubeObjectMenu toolbar gaps right bottom"
        data-testid="menu-actions-for-kube-object-menu-for-some-uid"
        id="menu-actions-for-kube-object-menu-for-some-uid"
      >
        <li>
          Some menu item
        </li>
        <li
          class="MenuItem"
          data-testid="menu-action-delete-for-/foo"
          tabindex="0"
        >
          <i
            class="Icon material interactive focusable"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="delete"
            >
              delete
            </span>
          </i>
          <div>
            Delete
          </div>
          <span
            class="title"
          >
            Delete
          </span>
        </li>
      </ul>
    </div>
  </div>
  <div
    class="Dialog flex center ConfirmDialog modal"
    data-testid="confirmation-dialog"
  >
    <div
      class="box"
    >
      <div
        class="confirm-content"
      >
        <i
          class="Icon material focusable big"
        >
          <span
            class="icon"
            data-icon-name="warning"
          >
            warning
          </span>
        </i>
         
        <p>
          Remove some-kind 
          <b>
            some-namespace/some-name
          </b>
           from 
          <b>
            some-context-name
          </b>
          ?
        </p>
      </div>
      <div
        class="confirm-buttons"
      >
        <button
          class="Button cancel plain"
          type="button"
        >
          Cancel
        </button>
        <button
          class="Button ok primary"
          data-testid="confirm"
          data-waiting="false"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`kube-object-menu given kube object without namespace when removing kube object, renders confirmation dialog without namespace 1`] = `
<body>
  <div>
    <div>
      <ul
        class="Menu MenuActions flex KubeObjectMenu toolbar gaps right bottom"
        data-testid="menu-actions-for-kube-object-menu-for-some-uid"
        id="menu-actions-for-kube-object-menu-for-some-uid"
      >
        <li>
          Some menu item
        </li>
        <li
          class="MenuItem"
          data-testid="menu-action-delete-for-/foo"
          tabindex="0"
        >
          <i
            class="Icon material interactive focusable"
            tabindex="0"
          >
            <span
              class="icon"
              data-icon-name="delete"
            >
              delete
            </span>
          </i>
          <div>
            Delete
          </div>
          <span
            class="title"
          >
            Delete
          </span>
        </li>
      </ul>
    </div>
  </div>
  <div
    class="Dialog flex center ConfirmDialog modal"
    data-testid="confirmation-dialog"
  >
    <div
      class="box"
    >
      <div
        class="confirm-content"
      >
        <i
          class="Icon material focusable big"
        >
          <span
            class="icon"
            data-icon-name="warning"
          >
            warning
          </span>
        </i>
         
        <p>
          Remove some-kind 
          <b>
            some-name
          </b>
           from 
          <b>
            some-context-name
          </b>
          ?
        </p>
      </div>
      <div
        class="confirm-buttons"
      >
        <button
          class="Button cancel plain"
          type="button"
        >
          Cancel
        </button>
        <button
          class="Button ok primary"
          data-testid="confirm"
          data-waiting="false"
          type="button"
        >
          Delete
        </button>
      </div>
    </div>
  </div>
</body>
`;

exports[`kube-object-menu given no kube object, renders 1`] = `
<body>
  <div>
    <ul
      class="Menu MenuActions flex KubeObjectMenu toolbar gaps right bottom"
      data-testid="menu-actions-for-kube-object-menu-for-undefined"
      id="menu-actions-for-kube-object-menu-for-undefined"
    />
  </div>
</body>
`;
